net/interfaces/defaultroute_ios.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build ios
+
+package interfaces
+
+import (
+	"log"
+)
+
+func defaultRoute() (d DefaultRouteDetails, err error) {
+	// We cannot rely on the delegated interface data on iOS. The NetworkExtension framework
+	// seems to set the delegate interface only once, upon the *creation* of the VPN tunnel.
+	// If a network transition (e.g. from Wi-Fi to Cellular) happens while the tunnel is
+	// connected, it will be ignored and we will still try to set Wi-Fi as the default route
+	// because the delegated interface is not updated by the NetworkExtension framework.
+	//
+	// Here we special-case iPhones and iPads with a simpler logic that doesn't look at
+	// the routing table: we try finding a hardcoded Wi-Fi interface, if it doesn't have
+	// an address, we fall back to cellular.
```

Comment:
This will work, but it's not a universal fix, and it doesn't address the edge cases (rare as those may be).

I mentioned using NetworkMonitor on the Swift side to simply re-bootstrap the VPN whenever we see a major interface change.   Keep track of the current/last interface, and just programatically call "stop" on NEPacketTunnelProvider if you see an interface show up that is cheaper than the one we're currently using (or the one you're currently using becomes unavailable).  This requires us to be always-on so that the system automatically restarts the tunnel however, which may not be the case on iOS.  This leaves the decision as to which interface to delegate to up to the OS.

Reply:
See refactor comment below. This suggestion gets the job done, however I am not a big fan of this approach as it would cause the VPN status bar item to appear, and I also imagine it could cause traffic leaks between the time the VPN goes down and it then gets re-established on the new interface.

I am now observing path updates on the Swift side, and I update the default route interface appropriately.